pkg/querier/worker/scheduler_processor_test.go

```diff
@@ -325,6 +327,63 @@ func TestCreateSchedulerProcessor(t *testing.T) {
 	assert.Equal(t, conf, sp.grpcConfig)
 }
 
+func TestSchedulerProcessor_ResponseStream(t *testing.T) {
```

Comment:
that's a neat test, I like the setup. Can you extend it a little bit? I'm thinking about cases where the query is cancelled midway sending or the frontend fails receiving the body. 

Also can you assert that the body is received intact - currently if the body is mangled (overwritten or a byte is skipped), the test will still pass.

Reply:
I found and fixed a problem with context cancellation (see 3603f80).
This also made me realize that the backoff/retry logic we currently have in place can't be used to retry a streaming query, because at the moment it would retry the whole sequence of requests (metadata and then the entire body). This would be rejected by the frontend.

For now I have opted to swallow any errors that happen during body streaming to not trigger a retry. Given that sending the metadata succeeded (retries will happen for this part) maybe it's ok to assume that we've got a healthy client and can stream out the response. Maybe however it would be better to keep some kind of streaming index and retry from there, wdyt?

Reply:
> Maybe however it would be better to keep some kind of streaming index and retry from there, wdyt?

I'm not quite sure to be honest. I don't know the failure modes of the query-frontend. I mean why would sending fail? If the query-frontend is dead, then there's no use of retrying. If the HTTP2 stream collapsed (or corrupted or got interrupted, idk), then how do we know what data the query-frontend has successfully received? Do we need to keep track of that in the QF end and rerequest it from the querier? At which point this sounds a lot like a TCP implementation.

IMO we can see actual failures first and then try to fix them. That's why I left a comment to surface the error message somehow so we can see what could go wrong and fix it then.

One middle ground is making the query-frontend tolerate restarting streams, so the querier can restart from the metadata and the first part, then the second part, etc. I don't know how much work that is